src/app/icd/client/RefreshKeySender.cpp

```diff
@@ -0,0 +1,144 @@
+                                   ICDClientStorage * aICDClientStorage) :
+    mOnConnectedCallback(HandleDeviceConnected, this),
+    mOnConnectionFailureCallback(HandleDeviceConnectionFailure, this)
+{
+    mpCheckInDelegate  = apCheckInDelegate;
+    mICDClientInfo     = aICDClientInfo;
+    mpICDClientStorage = aICDClientStorage;
+}
+
+void RefreshKeySender::Callback::OnResponse(chip::app::CommandSender * apCommandSender,
+                                            const chip::app::ConcreteCommandPath & aPath, const chip::app::StatusIB & aStatus,
+                                            chip::TLV::TLVReader * aData)
+{
+    app::Clusters::IcdManagement::Commands::RegisterClientResponse::DecodableType responseData;
+    if (aData == nullptr)
+    {
+        mpRefreshKeySender->GetCheckInDelegate()->OnKeyRefreshDone(mpRefreshKeySender, CHIP_ERROR_INVALID_TLV_ELEMENT);
+        return;
+    }
+
+    mError = DataModel::Decode(*aData, responseData);
+    if (mError != CHIP_NO_ERROR)
+    {
+        ChipLogError(ICD, "Failed to decode re-registration response data: %" CHIP_ERROR_FORMAT, mError.Format());
+        mpRefreshKeySender->GetCheckInDelegate()->OnKeyRefreshDone(mpRefreshKeySender, mError);
+        return;
+    }
+    // Update the ICDClientInfo with new key and start counter and store it to persistence
+    mpRefreshKeySender->GetICDClientInfo().start_icd_counter = responseData.ICDCounter;
+    mpRefreshKeySender->GetICDClientInfo().offset            = 0;
+    mpRefreshKeySender->GetICDClientStorage()->RemoveKey(mpRefreshKeySender->GetICDClientInfo());
+    mError = mpRefreshKeySender->GetICDClientStorage()->SetKey(mpRefreshKeySender->GetICDClientInfo(),
+                                                               mpRefreshKeySender->mNewKey.Span());
+    if (mError != CHIP_NO_ERROR)
+    {
+        ChipLogError(ICD, "Failed to set the new key after re-registration: %" CHIP_ERROR_FORMAT, mError.Format());
+        mpRefreshKeySender->GetCheckInDelegate()->OnKeyRefreshDone(mpRefreshKeySender, mError);
+        return;
+    }
+
+    mError = mpRefreshKeySender->GetICDClientStorage()->StoreEntry(mpRefreshKeySender->GetICDClientInfo());
+    if (mError != CHIP_NO_ERROR)
+    {
+        ChipLogError(ICD, "Failed to store the new key after re-registration: %" CHIP_ERROR_FORMAT, mError.Format());
+        mpRefreshKeySender->GetCheckInDelegate()->OnKeyRefreshDone(mpRefreshKeySender, mError);
+        return;
+    }
+
+    mpRefreshKeySender->GetCheckInDelegate()->OnCheckInComplete(mpRefreshKeySender->GetICDClientInfo());
+    mpRefreshKeySender->GetCheckInDelegate()->OnKeyRefreshDone(mpRefreshKeySender, CHIP_NO_ERROR);
+}
+void RefreshKeySender::Callback::OnError(const chip::app::CommandSender * apCommandSender, CHIP_ERROR aError)
+{
+    mError = aError;
+    mpRefreshKeySender->GetCheckInDelegate()->OnKeyRefreshDone(mpRefreshKeySender, mError);
+}
```

Comment:
I couldn't initialize mNewKey in the initializer list because of this error: implicitly-declared ‘constexpr chip::Crypto::SensitiveDataBuffer<16>::SensitiveDataBuffer(const chip::Crypto::SensitiveDataBuffer<16>&)’ is deprecated [-Werror=deprecated-copy]

Reply:
Follow up issue created